evallib/taskqueue.py

```diff
@@ -0,0 +1,129 @@
+        # Whether to show a progress bar on stdout
+        self.show_progress_bar = show_progress_bar
+
+    def run_tasks(
+        self,
+        tasks: List[T],
+        *,
+        run: Callable[[T], Optional[str]],
+        output_file: Optional[str] = None,
+    ):
+        """
+        `run_tasks` is a light-weight, static distributed task-queue.
+
+        Given a list `tasks` of task descriptions of a generic type `T`, and a function
+        `run` which takes a task description as input and executes the task, `run_tasks`
+        will run all tasks, in a parallelized way.
+
+        It is static, meaning that the list of all tasks to be run needs to be known
+        entirely up-front, which allows us to distribute the tasks over many machines
+        in a straightforward way, without requiring complex coordination.
+
+        The `config` controls the parallelism, where `config.max_threads` controls
+        the degree of multi-threading for a single invocation of the script, and
+        `self.stride` and `self.offset` enable distribution over multiple invocations
+        (and hence, over multiple machines), where
+        * `self.stride` should be set to the total number of parallel jobs being run, and
+        * Each parallel job should get a different value of `self.offset`, ranging from
+            0 to `self.stride-1` (inclusive), to guarantee each task is run exactly once.
+
+        Note that, since tasks may be distributed over multiple invocations of a script,
+        the `run` function should *not* rely on mutating global state to aggregate results.
+        Instead, `run` is expected to return a single line (without terminating new-line)
+        in some format which is safe to concatenate (preferably jsonl), the output of each task 
+        (run by the current job) is concatenated and collected in a single file at the path
+        specified by `output_file`. Note that this file is truncated if it already
+        exists. If `output_file` is None, output is printed to stdout instead.
+
+        If `run` returns `None` for a task, no output is produced (in particular, 
+        we do *not* write an empty line in such a case).
+        """
+
+        tasks = tasks[self.offset::self.stride]
+
+        with ThreadPoolExecutor(max_workers=self.max_threads) as executor:
+            futures = [executor.submit(run, task) for task in tasks]
+            if self.show_progress_bar:
+                futures = tqdm(futures)
+
+            if output_file is None:
+                for future in futures:
+                    result = future.result()
+                    if result is not None:
+                        print(result)
+            else:
+                with open(output_file, "w") as f_out:
+                    for future in futures:
+                        result = future.result()
+                        if result is not None:
+                            f_out.write(str(result) + '\n')
```

Comment:
always use `f_out` and set `f_out` to `stdout` instead of duplicating this logic.